block-streamer/src/receiver.rs

```diff
@@ -0,0 +1,259 @@
+            .or_insert(vec![chunk.clone()]);
+        Ok(())
+    }
+
+    pub fn handle_cid_msg(&mut self, cid: Vec<u8>) -> Result<()> {
+        let cid_marker = &cid[..10];
+        self.cids_to_build.entry(cid_marker.to_vec()).or_insert(cid);
+        Ok(())
+    }
+
+    pub fn attempt_block_assembly(&mut self) -> Result<()> {
+        for (cid_marker, cid) in &self.cids_to_build {
+            if let Some(cid_chunks) = &self.cid_chunks.get(cid_marker) {
+                if let Ok(wrapper) = BlockWrapper::from_chunks(cid, cid_chunks) {
+                    let block = wrapper.to_block()?;
+                    if !block.links().is_empty() {
+                        info!(
+                            "Found root block {} with links {:?}",
+                            &block.cid(),
+                            &block.links()
+                        );
+                        self.roots.insert(*block.cid(), block);
+                    } else {
+                        info!(
+                            "Found child block {} with {} bytes",
+                            &block.cid(),
+                            block.data().len()
+                        );
+                        self.blocks.insert(*block.cid(), block.clone());
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn attempt_tree_assembly(&self) -> Result<()> {
+        // TODO could optimize this to only attempt block assembly for the last seen CID
+        for (_, root) in self.roots.iter() {
+            if self.verify_block_complete(*root.cid()).unwrap() {
+                info!("Block complete: {}", &root.cid());
+                let path = PathBuf::from(root.cid().to_string());
+                if chunks_to_path(&path, root, &self.blocks).await? {
+                    info!("Assembly success! {}", &path.display());
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn verify_block_complete(&self, cid: Cid) -> Result<bool> {
+        let block = self.roots.get(&cid);
+        match block {
+            Some(block) => {
+                // Check if all child blocks are present
+                for c in block.links().iter() {
+                    if !self.blocks.contains_key(c) {
+                        info!("Missing cid {}, wait for more data", c);
+                        return Ok(false);
+                    }
+                }
+                Ok(true)
+            }
+            None => {
+                info!("Failed to find block for {}", cid);
+                Ok(false)
+            }
+        }
+    }
+
+    pub async fn handle_transmission_msg(&mut self, msg: TransmissionMessage) -> Result<()> {
+        match msg {
+            TransmissionMessage::Chunk(chunk) => self.handle_chunk_msg(chunk)?,
+            TransmissionMessage::Cid(cid) => self.handle_cid_msg(cid)?,
+        }
+        self.attempt_block_assembly()?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
```

Comment:
Nice to see the added tests 💯   We may want to add a light CI in the near future for running tests and linting.